[OSDOCS-3059] ShiftStack installation with OVS-DPDK workers (#41484)
openshift/openshift-docs: installing/installing_openstack/installing-openstack-installer-ovs-dpdk.adoc

// TODO Duplicate for DPDK

include::modules/networking-osp-preparing-for-sr-iov.adoc[leveloffset=+1]
include::modules/networking-osp-enabling-metadata.adoc[leveloffset=+2]
include::modules/networking-osp-enabling-vfio-noiommu.adoc[leveloffset=+2]
include::modules/installation-osp-dpdk-binding-vfio-pci.adoc[leveloffset=+2]
include::modules/installation-osp-dpdk-exposing-host-interface.adoc[leveloffset=+2]

// TODO: Verify me
The cluster is installed and prepared for configuration. You must now perform the SR-IOV and OVS-DPDK configuration tasks in "Next steps".

include::modules/cluster-telemetry.adoc[leveloffset=+1]

.Additional resources

* See xref:../../support/remote_health_monitoring/about-remote-health-monitoring.adoc#about-remote-health-monitoring[About remote health monitoring] for more information about the Telemetry service

== Additional resources
* See xref:../../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-understanding-low-latency_cnf-master[Performance Addon Operator for low latency nodes] for information about configuring your deployment for real-time running and low latency.

== Next steps
// TODO: verify machineset

> * To complete SR-IOV configuration for your cluster:

Author: ```suggestion
// TODO: Write up proper next steps
* To complete SR-IOV configuration for your cluster:
```

== Prerequisites

* Review details about the
xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update]
processes.
** Verify that {product-title} {product-version} is compatible with your {rh-openstack} version by using the "Supported platforms for OpenShift clusters" section. You can also compare platform support across different versions by viewing the link:https://access.redhat.com/articles/4679401[{product-title} on {rh-openstack} support matrix].

* Have a storage service installed in {rh-openstack}, like block storage (Cinder) or object storage (Swift). Object storage is the recommended storage technology for {product-title} registry cluster deployment. For more information, see xref:../../scalability_and_performance/optimizing-storage.adoc#optimizing-storage[Optimizing storage].

* Have the metadata service enabled in {rh-openstack}.

> * Plan your {rh-openstack} OVS-DPDK deployment by referring to link:https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.0/html/network_functions_virtualization_planning_and_configuration_guide/assembly_ovsdpdk_parameters[Planning your OVS-DPDK deployment] in the Network Functions Virtualization Planning and Configuration Guide.

Author: ```suggestion
* Plan your {rh-openstack} OVS-DPDK deployment by referring to link:https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.2/html/network_functions_virtualization_planning_and_configuration_guide/assembly_ovsdpdk_parameters[Planning your OVS-DPDK deployment] in the Network Functions Virtualization Planning and Configuration Guide.
```

> * Configure your {rh-openstack} OVS-DPDK deployment according to link:https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.0/html/network_functions_virtualization_planning_and_configuration_guide/part-dpdk-configure[Configuring an OVS-DPDK deployment] in the Network Functions Virtualization Planning and Configuration Guide.

Author: ```suggestion
* Configure your {rh-openstack} OVS-DPDK deployment according to link:https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.2/html/network_functions_virtualization_planning_and_configuration_guide/part-dpdk-configure[Configuring an OVS-DPDK deployment] in the Network Functions Virtualization Planning and Configuration Guide.

** You must complete link:https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.2/html/network_functions_virtualization_planning_and_configuration_guide/part-dpdk-configure#p-ovs-dpdk-flavor-deploy-instance["Creating a flavor and deploying an instance for OVS-DPDK"] before you begin this process. 
```

:_content-type: ASSEMBLY
[id="installing-openstack-installer-ovs-dpdk"]
= Installing a cluster on OpenStack that supports DPDK-connected compute machines
include::modules/common-attributes.adoc[]
:context: installing-openstack-installer-ovs-dpdk
:FeatureName: Installing a cluster on {rh-openstack} that supports DPDK-connected compute machines

> toc::[]

Author: lgtm now. Feel free to take a peek if you like.

> toc::[]

Reviewer: I'm still not seeing the TOC rendered in the [preview URL](https://deploy-preview-41484--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-ovs-dpdk.html) but that could be because I'm not very good at tracking multiple commits. 

![image](https://user-images.githubusercontent.com/47002677/157118655-dadb1610-8a0f-4258-a18d-b9359f25d2e6.png)

If you say it's good now, I believe it.

> toc::[]

Author: @bobfuru Let's get maddeningly specific about this, as I'm curious and you're providing a screenshot. 

I'm not seeing anything on my preview that is different from the layout on prod. My brain is fried from trying to potty train my toddler over the last few days, so I'm not sure if I'm missing something really obvious. :) 

Could you mark up on the screenshot where you think the secondary TOC should be?

> toc::[]

Reviewer: I owe you a great big apology for my terrible lie, @maxwelldb. For no good reason, I was looking for the old school inline TOC that we had before Aidan gave us the righthand TOC. What you have is absolutely splendid. I'm sorry for wasting your time with this.  

My brain is still fried from potty training my toddlers and it's been years. Best of luck! And thanks for your patience with my feedback.

> toc::[]

Author: @bobfuru No worries. I just wanted to make sure I hadn't gone UI blind or something. Thanks!

include::modules/installation-osp-accessing-api.adoc[leveloffset=+1]
include::modules/installation-osp-accessing-api-floating.adoc[leveloffset=+2]
include::modules/installation-osp-accessing-api-no-floating.adoc[leveloffset=+2]
include::modules/installation-osp-configuring-sr-iov.adoc[leveloffset=+1]
include::modules/installation-launching-installer.adoc[leveloffset=+1]
include::modules/installation-osp-verifying-cluster-status.adoc[leveloffset=+1]
include::modules/cli-logging-in-kubeadmin.adoc[leveloffset=+1]

The cluster is operational. Before you can add OVS-DPDK compute machines, though, you must perform additional tasks.

include::modules/networking-osp-enabling-metadata.adoc[leveloffset=+1]
include::modules/networking-osp-enabling-vfio-noiommu.adoc[leveloffset=+1]
include::modules/installation-osp-dpdk-binding-vfio-pci.adoc[leveloffset=+1]
include::modules/installation-osp-dpdk-exposing-host-interface.adoc[leveloffset=+1]

.Additional resources

* xref:../../networking/multiple_networks/configuring-additional-network.adoc#nw-multus-host-device-object_configuring-additional-network[Creating an additional network attachment with the Cluster Network Operator]

> The cluster is installed and prepared for configuration. You must now perform the OVS-DPDK configuration tasks in "Next steps".

Author: ```suggestion
The cluster is installed and prepared for configuration. You must now perform the OVS-DPDK configuration tasks in <<next-steps_installing-openstack-installer-ovs-dpdk, "Next steps">>.
```

include::modules/cluster-telemetry.adoc[leveloffset=+1]

.Additional resources

* See xref:../../support/remote_health_monitoring/about-remote-health-monitoring.adoc#about-remote-health-monitoring[About remote health monitoring] for more information about the Telemetry service

> == Additional resources
> * See xref:../../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-understanding-low-latency_cnf-master[Performance Addon Operator for low latency nodes] for information about configuring your deployment for real-time running and low latency.
> 
> == Next steps

Author: ```suggestion
[id="additional-resources_installing-openstack-installer-ovs-dpdk"]
== Additional resources
* See xref:../../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-understanding-low-latency_cnf-master[Performance Addon Operator for low latency nodes] for information about configuring your deployment for real-time running and low latency.

[id="next-steps_installing-openstack-installer-ovs-dpdk"]
== Next steps
```

include::modules/installation-configuration-parameters.adoc[leveloffset=+1]
include::modules/installation-osp-custom-subnet.adoc[leveloffset=+2]
include::modules/installation-osp-deploying-bare-metal-machines.adoc[leveloffset=+2]
include::modules/installation-osp-config-yaml.adoc[leveloffset=+2]
include::modules/ssh-agent-using.adoc[leveloffset=+1]
include::modules/installation-osp-accessing-api.adoc[leveloffset=+1]
include::modules/installation-osp-accessing-api-floating.adoc[leveloffset=+2]
include::modules/installation-osp-accessing-api-no-floating.adoc[leveloffset=+2]
include::modules/installation-osp-configuring-sr-iov.adoc[leveloffset=+1]
include::modules/installation-launching-installer.adoc[leveloffset=+1]
include::modules/installation-osp-verifying-cluster-status.adoc[leveloffset=+1]
include::modules/cli-logging-in-kubeadmin.adoc[leveloffset=+1]

The cluster is operational. Before you can add OVS-DPDK compute machines, though, you must perform additional tasks.

// TODO Duplicate required mods/language for DPDK and cut what's not necessary

// include::modules/networking-osp-preparing-for-sr-iov.adoc[leveloffset=+1]

> // TODO: Metadata required or not?

Reviewer 2: config-drive or metadata are required. Note that all Red Hat OpenStack clouds provide metadata service, so no worries.

Author: ```suggestion
```

:_content-type: ASSEMBLY
[id="installing-openstack-installer-ovs-dpdk"]
= Installing a cluster on OpenStack that supports DPDK-connected compute machines
include::modules/common-attributes.adoc[]
:context: installing-openstack-installer-ovs-dpdk

> :FeatureName: Installing a cluster on {rh-openstack} that supports DPDK-connected compute machines

Author: ```suggestion
```

include::modules/installation-osp-accessing-api.adoc[leveloffset=+1]
include::modules/installation-osp-accessing-api-floating.adoc[leveloffset=+2]
include::modules/installation-osp-accessing-api-no-floating.adoc[leveloffset=+2]
include::modules/installation-osp-configuring-sr-iov.adoc[leveloffset=+1]
include::modules/installation-launching-installer.adoc[leveloffset=+1]
include::modules/installation-osp-verifying-cluster-status.adoc[leveloffset=+1]
include::modules/cli-logging-in-kubeadmin.adoc[leveloffset=+1]

The cluster is operational. Before you can add OVS-DPDK compute machines though, you must perform additional tasks.

include::modules/networking-osp-enabling-metadata.adoc[leveloffset=+1]
include::modules/networking-osp-enabling-vfio-noiommu.adoc[leveloffset=+1]
include::modules/installation-osp-dpdk-binding-vfio-pci.adoc[leveloffset=+1]
include::modules/installation-osp-dpdk-exposing-host-interface.adoc[leveloffset=+1]

.Additional resources

* xref:../../networking/multiple_networks/configuring-additional-network.adoc#nw-multus-host-device-object_configuring-additional-network[Creating an additional network attachment with the Cluster Network Operator]

> The cluster is installed and prepared for configuration. You must now perform the OVS-DPDK configuration tasks in <<next-steps_installing-openstack-installer-ovs-dpdk, "Next steps">>.

Author: ```suggestion
The cluster is installed and prepared for configuration. You must now perform the OVS-DPDK configuration tasks in <<next-steps_installing-openstack-installer-ovs-dpdk, Next steps>>.
```
